### PR TITLE
[RFC] CSS property order

### PR DESCRIPTION
**Pull request by kaelig**
Let's explore together a way to have a property order that looks satisfying.

I've added a bunch of properties that were missing, put `width` before `height`, and moved `transition` close to `transform` (to be discussed).

### The Plan

- [x] Discuss the shortcomings of the current order
- [x] Decide what is worth changing (we don't want to break everyone's workflow)
- [ ] ~Document decisions in comment blocks (e.g.: why we omitted certain properties, why top/right/bottom/left and not top/bottom/left/right?)~
- [x] Upgrade rule syntax to the latest version of stylelint-order

This is a safe space. A lot of us care deeply about this, so be careful and avoid to get too worked up so we can keep this civil :) Please try to defend your opinions with a bit of context.

cc @Shopify/fed

### CODE REVIEW

**Comment by kaelig** [comment]
I like this from @cshold:

> What is common is we want to write CSS quickly in an organized way

Maybe we could enforce "grouping" without enforcing order. The downside is that gzip compression suffers a bit when the order isn't strict (not sure that's a concern though). Thoughts?

File: rules/order.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
 module.exports = {
   'order/properties-order': [
     {
       order: 'strict',
       properties: [
         'content',
+        'quotes',
       ],
     },
     {
```
Comment (dan-menard) [inline rules/order.js]: Putting `z-index` between `position` and the things that matter to position (top/right/bottom/left) seems strange to me.

File: rules/order.js [hunk 2]
```diff
@@ -15,13 +18,35 @@ module.exports = {
...
+        'visibility',
+        'opacity',
       ],
     },
     {
       order: 'strict',
       properties: [
         'transform',
         'transform-origin',
+        // Animating transforms is common, so let's group them together
```
Comment (dan-menard) [inline rules/order.js]: ❤️

File: rules/order.js (hunk 1 above)
Comment (kaelig) [inline rules/order.js]: Good to know, I guess it could be grouped with visibility/opacity?

File: rules/order.js [hunk 3]
```diff
@@ -30,6 +55,7 @@ module.exports = {
         'box-sizing',
         'display',
         'overflow',
+        'vertical-align',
```
Comment (dan-menard) [inline rules/order.js]: This one, and even more so `overflow` seem misplaced to me. Two reasons for this:

1. Like my comment about `position`, it seems odd to put anything between `display` and display-based properties, like the `flex-` stuff.

2. Anecdotally, I generally treat `vertical-align` and especially `overflow` as an after-thought, like I write up a nice little layout that works for the common case, and add `overflow` to fix an edge case or `vertical-align` to get the spacing just right. Because of that, I'd rather have these pinned to the bottom than so close to the top.

Not a strong stance here, I'm curious to hear others' thoughts :)

File: rules/order.js [hunk 4]
```diff
@@ -41,8 +67,30 @@ module.exports = {
...
+        'grid-column-gap',
+        'grid-row-gap',
+        'columns',
+        'column-gap',
+        'column-fill',
+        'column-rule',
+        'column-span',
+        'column-count',
+        'column-width',
+        'float',
```
Comment (dan-menard) [inline rules/order.js]: `float` (and I guess `clear`) is interesting, because we kind of want to discourage them. Is there any way to hint that here? Does putting them at the very bottom imply they're less useful than things at the top, for instance?

Separately, have we considered (or do we already have) a "no-floats" rule?

File: rules/order.js [hunk 5]
```diff
@@ -59,18 +107,59 @@ module.exports = {
...
         'background',
+        'background-attachment',
+        'background-clip',
         'background-color',
-        'background-size',
-        'background-position',
+        'background-image',
         'background-repeat',
+        'background-position',
+        'background-size',
```
Comment (dan-menard) [inline rules/order.js]: The background stuff is all alphabetical except that -repeat and -position are flipped. Is it meant to be alphabetical?

If not, I notice you swapped the order of -position and -size. What's the rationale there? (or was it unintentional?)

File: rules/order.js [hunk 6]
```diff
@@ -59,18 +107,59 @@ module.exports = {
...
+        'border-style',
+        'border-top-style',
+        'border-right-style',
+        'border-bottom-style',
+        'border-left-style',
+        'border-radius',
+        'border-top-left-radius',
+        'border-top-right-radius',
+        'border-bottom-left-radius',
+        'border-bottom-right-radius',
```
Comment (dan-menard) [inline rules/order.js]:
```
'border-top-left-radius',
'border-top-right-radius',
'border-bottom-left-radius',
'border-bottom-right-radius',
```

^ This feels inconsistent to me, because the `border-width` and `-style` properties follow a clockwise rotation (top right bottom left) and these don't because `bottom-left` is before `bottom-right`.

File: rules/order.js (hunk 1 above)
Comment (beefchimi) [inline rules/order.js]: I would actually expect `z-index` to immediately follow a `position` declaration. Sure there are several properties that create a stacking context - but `position` is the main one we all think of.

I'm not super opinionated about this tho.

File: rules/order.js [hunk 7]
```diff
@@ -85,15 +174,23 @@ module.exports = {
         'line-height',
         'color',
         'text-align',
-        'white-space',
-        'vertical-align',
```
Comment (dan-menard) [inline rules/order.js]: It's a little strange to me that `vertical-align` was moved, but `line-height` wasn't. Those two are highly related, and should probably live together (either here or up in the higher block).

File: rules/order.js (hunk 3 above)
Comment (beefchimi) [inline rules/order.js]: I have no strong stance either but lets be real, `vertical-align` is the most useless property. How often do we even use it?

File: rules/order.js (hunk 5 above)
Comment (beefchimi) [inline rules/order.js]: Isn't there an official order to these properties when using shorthand?

File: rules/order.js [hunk 8]
```diff
@@ -85,15 +174,23 @@ module.exports = {
...
+        'text-transform',
         'text-decoration',
+        'text-rendering',
+        'text-shadow',
+        'text-overflow',
+        'letter-spacing',
+        'word-spacing',
+        'word-wrap',
+        'word-break',
+        'white-space',
```
Comment (dan-menard) [inline rules/order.js]: Might make sense to move `overflow` down here, actually, you almost always want to set it when setting `white-space` an/or `text-overflow`.

**Comment by cshold** [comment]
I'm probably not the best person to comment on the order because to me it's not solving CSS organization problems that we face. Whether `height` is before or after `width` doesn't effect how I read people's styles, nor does it impact the quality of their work.

I do like grouping similar properties together, but I'd argue we hire experienced enough people that we'll never see something like this:
```
.block {
  padding: 10px;
  font-style: italic;
  @include transition-mixin;
  align-items: center;
  margin: 0 10px;
  display: flexbox;
}
```
If we see that in a PR, a comment educating the contributor about grouping is satisfactory IMO. That said I've never seen a PR like that.

Where I think the breakdown in workflow happens is when we dictate what _has_ to be above/below other properties. Take these two examples:

##### Example 1
```
.block {
  display: block;
  padding: 20px;
  margin: 0 10px;
  font-style: italic;
  font-weight: bold;
  color: blue;
}
```

##### Example 2
```
.block {
  color: blue;
  font-style: italic;
  font-weight: bold;
  display: block;
  margin: 0 10px;
  padding: 20px;
}
```

Which one is better? Why? 

Linting the property order is us definitively saying one is wrong.

The bigger issues that we face in writing organized CSS, some of which are already linted, are:
- proper use and consistency in selector naming conventions (BEM, or the new Polaris styles with capital letters and auto-generated class names)
- selector length
- selector nesting — aka specificity

File: rules/order.js (hunk 5 above)
Comment (dan-menard) [inline rules/order.js]: ooh! Good thinking: https://developer.mozilla.org/en/docs/Web/CSS/background

According to that, we missed `background-origin` (should be directly after `-clip`) and `-position` and `-repeat` are backwards.

**Comment by dan-menard** [comment]
Like Carson, I don't like the idea of enforcing order. I agree with his reason above, but want to add one more:

Every time we add something like this, we're adding one more thing to worry about when writing code. It may not seem catastrophic in isolation, but the more we add tools and validation like this, the harder it becomes to accept the tradeoff: that the consistency it buys is worth the overhead it adds.

I believe there's a real sanity cost to linting little details like this, because there's [real, actual science](http://work.chron.com/effects-autonomy-job-satisfaction-14677.html) about the benefits of autonomy. And in this case, I just don't think the bonus we gain from consistency is worthwhile.

File: rules/order.js (hunk 3 above)
Comment (dleroux) [inline rules/order.js]: I personally like overflow after width and height, flows logically after specifying dimensions.

**Comment by m-ux** [comment]
IMO, when it comes to CSS - the things I care about (from an organizational perspective) is clarity and consistency around the hierarchy of elements - not the order in which properties of said elements appear.  

I've never fully understood this, but can anyone help me understand what the perceived value of a consistent order of properties is?  I see more detriment in having to memorize this list, accidentally making a mistake, getting flagged by a linter and having to spend extra time manually fixing the order of properties than I see value in ensuring the props always occur in the proper organized order.

Does this simple satisfy aesthetic pleasure at reading a clean consistent file, or is there benefit beyond this?

**Comment by harisaurus** [comment]
I agree with Carson. Giving
```
'border-top-left-radius',
'border-top-right-radius',
'border-bottom-left-radius',
'border-bottom-right-radius',
```
a specific order, for example, is unnecessary as long as they're grouped together. Same goes for flex properties, background properties etc.

File: rules/order.js (hunk 2 above)
Comment (lemonmade) [inline rules/order.js]: IMO this actually isn't really true, we often also transition opacity and, less frequently, background-color. Always has felt right to me for these to be at the bottom.

File: rules/order.js [hunk 9]
```diff
@@ -1,9 +1,12 @@
+// Order inspired by Concentric CSS
+// http://rhodesmill.org/brandon/2011/concentric-css/
 module.exports = {
   'order/properties-order': [
     {
       order: 'strict',
       properties: [
         'content',
+        'quotes',
```
Comment (lemonmade) [inline rules/order.js]: Why have this up here? Seems typographic in nature, no?

**Comment by dan-menard** [comment]
> if the order were not strict but flexible within the groups, would that be better, or worse?

Sounds better to me 👍

I think the current system is doing good things, it just feels too restrictive. Ideally I'd want something where:

1. This is valid (width and height are grouped, background properties are grouped):

```
.myClass {
  width: 1080px;
  height: 720px;
  background-color: color(sky);
  background-image: url('admin/clouds.svg');
```

2. And this is valid (same grouping, different order within the group, and within the selector):

```
.myClass {
  background-image: url('admin/clouds.svg');
  background-color: color(sky);
  height: 720px;
  width: 1080px;
```

3. But this is invalid (groups are mixed):

```
.myClass {
  width: 1080px;
  background-color: color(sky);
  height: 720px;
  background-image: url('admin/clouds.svg');
```

I like this because it enforces the thing we all seem to agree on (group like properties) without enforcing the things that divide us (strict order for groups, strict order within groups).

That said, I have no idea if arbitrary order for groups is possible, so if the best we can do is arbitrary order within groups, I'll take it.

**Comment by beefchimi** [comment]
In the CSS I author for myself - I'm strict about order. In the CSS I work on within a company, I think grouping _probably_ works best – and its a step forward.

File: rules/order.js (hunk 2 above)
Comment (kaelig) [inline rules/order.js]: I was thinking opacity could be in this group as well.

Not sure we should put them at the bottom though. Transforms have a big impact on the positioning and in Concentric CSS fashion it would probably make sense to have those at the top.

File: rules/order.js (hunk 3 above)
Comment (gavinsmith) [inline rules/order.js]: This sort of stuff gives me the heebie jeebies about this whole thing. The order is often so, so subjective, and there are hundreds of properties. It's an unreasonable ask, IMO, that _everyone_ actually get on board with an at least somewhat arbitrary grouping when there is absolutely not necessarily one 'right way' form a logical standpoint. This will require back and forth with the linter in one way or another for the majority of contributors, there is no way around it in my view. That is what we lose.

File: rules/order.js (hunk 2 above)
Comment (lemonmade) [inline rules/order.js]: (I was referring to transitions at the bottom, not transforms, btw)

File: rules/order.js (hunk 3 above)
Comment (lemonmade) [inline rules/order.js]: Aren't all stylistic rules inherently arbitrary, and don't they all cause back-and-forth with the linter for people who don't know them? Are we arguing that we should never enforce any consistent styling anywhere?

Comment (gavinsmith) [inline rules/order.js]: I don't think it's a fair comparison, most of our code style guidelines (and linting) is stuff that can be learned and remembered pretty easily. You can also often look at surrounding code to find out what the norm is. Neither of those is that simple with CSS props because there are so many of them, and finding one other instance of the prop you're trying to place does not give you all of the information you need to decide precisely where it needs to go to satisfy a linter. You're guessing or reading a detailed style 'spec' for it.

IMO, this is more akin to enforcing ordering groupings and classes in a ruby large helper file based on what type of functionality that helper helps with. Sometimes it's murky and I anticipate there will be a _lot_ of guessing.

I wrote up a car metaphor and then deleted it because that's dumb, but to many, a lot of CSS props are like a [ute](http://s3.caradvice.com.au/thumb/770/382/wp-content/uploads/2015/01/2015-holdenvsford-VFFGX-ute-comparison-23.jpg) in that their categorized purpose is not objective and could easily be viewed multiple ways.

Anyway, sorry for re-booting the original debate inside this arbitrary thread, but I personally don't think it's done with.

File: rules/order.js (hunk 2 above)
Comment (kaelig) [inline rules/order.js]: Oh, gotcha.

**Comment by AWaselnuk** [comment]
Enforcing some sort of grouping order makes the most sense to me. That is something that can be quickly learned and memorized and help any developer understand the role of CSS properties and what level they are working at in the code base. A grouping by order is also enough to quickly find exactly what you need when reading code.

Stricter order than that (ie. `height` has to follow `width`) seems almost analogous to asking people to strictly order the JavaScript variables they declare at the top of a method. It feels like the wrong level of abstraction, and feels like a major memorization task

**Comment by m-ux** [comment]
I'm still skeptical about the benefit, but have no qualms with defining and highlighting property groupings (I have done this in the past as well).  I really like Adam's point re: `help any developer understand the role of CSS properties and what level they are working at in the code base`.  

What are some of the proposed groupings? How should we go about highlighting these groups within a css file?

**Comment by kaelig** [comment]
It looks like there's an unanimous preference towards grouping rather than strict ordering.
Not sure how to highlight said groups within a CSS file - or if that's needed at all.

For the groupings, how about those ones? Note: properties wouldn't need to be in a particular order.

```css
.selector {
  /* positioning */
  position
  z-index
  top
  right
  bottom
  left
  visibility
  opacity

  /* box-model */
  box-sizing
  display
  overflow
  vertical-align
  flex
  flex-grow
  flex-shrink
  flex-basis
  order
  align-self
  flex-direction
  flex-wrap
  justify-content
  align-items
  align-content
  grid
  grid-template-rows
  grid-template-columns
  grid-template-areas
  grid-auto-rows
  grid-auto-columns
  grid-auto-flow
  grid-column-gap
  grid-row-gap
  columns
  column-gap
  column-fill
  column-rule
  column-span
  column-count
  column-width
  float
  clear
  width
  min-width
  max-width
  height
  min-height
  max-height
  margin
  margin-top
  margin-right
  margin-bottom
  margin-left
  padding
  padding-top
  padding-right
  padding-bottom
  padding-left
  border
  border-top
  border-right
  border-bottom
  border-left
  border-width
  border-top-width
  border-right-width
  border-bottom-width
  border-left-width
  border-style
  border-top-style
  border-right-style
  border-bottom-style
  border-left-style
  border-radius
  border-top-left-radius
  border-top-right-radius
  border-bottom-left-radius
  border-bottom-right-radius
  border-color
  border-top-color
  border-right-color
  border-bottom-color
  border-left-color
  outline
  box-shadow
  
  /* Others */
}
```

**Comment by lemonmade** [comment]
I am 👍 for the broad grouping, no strict ordering within. I would propose these groupings, which is roughly the basis of what we have now based on my perception of our standards when I wrote the config originally:

```
some weirdo properties: content;
positioning: position, top, bottom, left, right, etc;
box model: display, padding, margin, flex, grid, etc;
style: background, box-shadow, border, visibility;
typography: font-family, line-height, color, etc;
```

Transform and transition probably have to be allowed to float since they don't fit neatly into a particular category.

**Comment by kaelig** [comment]
Thanks @lemonmade, I updated the list above.

A few notes:

- `border` sort of has an impact on the box model so I'd be in favor of keeping it grouped with other box model properties
- `visibility` is a key component of the mental model you draw when reading a CSS ruleset. I propose keeping it at the top with positioning & visibility.
- weird props like `content`: good group to have! Added it to the list above.
- style & typography: we agree they should come after positioning & box-model, but I'd be in favor of keeping the order pretty loose after those two groups so that it's easy to enforce. Does that sound fair?

**Comment by lemonmade** [comment]
I don't really get `visibility` being separate from `opacity`, nor how it relates to the box model in the non-table case. I feel that typography is a pretty distinct category and makes sense to be separated, but I don't care too much about that one.

**Comment by gavinsmith** [comment]
Not to sound like a broken record, but these ambiguities are precisely my concern. If we have to discuss at length how to classify various properties, how is anybody else to stand a chance?

**Comment by lemonmade** [comment]
As has been brought up many times, listing rules often end up being this way, and the linter's auto fixing/ error messages are what makes this not as much of a burden on the actual work

**Comment by kaelig** [comment]
I took into account people's feelings and came up with an order that's a fair compromise between readability and ease of authoring.

I documented this order in the file, see https://github.com/Shopify/stylelint-config-shopify/blob/property-order/rules/order.js

### tl;dr

- Put variables & custom properties at the top (unless anyone feels strongly about this)
- Then come weird props, positioning & box model properties
- All other properties come after
- No specific property order is enforced

### The following patterns are _not_ considered warnings:

```scss
.Foo {
  $foo: 'foo';
  position: relative;
  display: block;
  margin: 10px;
  color: $foo;
}

.Foo {
  $foo: 'foo';
  position: relative;
  margin: 10px;
  display: block;
  color: $foo;
}
```

### The following patterns are considered warnings:

```scss
.Foo {
  position: relative;
  display: block;
  $foo: 'foo';
  color: $foo;
}

.Foo {
  $foo: 'foo';
  color: $foo;
  position: relative;
  display: block;
}
```

**Comment by kaelig** [comment]
Hey everyone - I just released `v3.0.0-beta.3` (Circle CI failed for betas 1 and 2…), you can test this set of rules in your projects using `yarn add stylelint-config-shopify@next --dev`.
